### DISCUSSION TITLE
AutoTP for training?

### DISCUSSION THREAD

**Discussion by KaiLv69** [deepspeedai/DeepSpeed#4930] [General]
Hi. I've noticed that DeepSpeed supports AutoTP for many models in HF. That's really useful. And I'm wondering whether it could be used to training models with TP. That's to say, can we do backpropagation with AutoTP?

**Comment by delock** [comment]
AutoTP relies on `LinarAllreduce`, `LmHeadLinearAllreduce`, `Linear Layer` in the files below.  They don't have `backward` function yet.
https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/module_inject/layers.py

**Reply by RezaYazdaniAminabadi** [reply]
I think it should be doable, and as @delock suggested we need some backward functions added for several operations that parallelize the model-execution. In addition, we need some high-level apis to manage the training environment such as data-loader, sharding optimizer states, ckpt load/saving, etc. Most of these are already handled through the Megatron-DeepSpeed repo, however, it only support megatron-type of models. If we can borrow some of those logic and build the core to handle tensor-parallelism then we can support more models from HF or other libraries.

**Reply by KaiLv69** [reply]
Great! Is there any plan to support that?

**Comment by delock** [comment]
I see this PR addressing this discussion https://github.com/microsoft/DeepSpeed/pull/5445

**Comment by delock** [comment]
@KaiLv69  AutoTP training had been supported in DeepSpeed.   https://github.com/deepspeedai/DeepSpeed/pull/7004  Thanks @inkcherry for contributing this feature!

**Reply by KaiLv69** [reply]
awesome!
